File: 133-clone-graph/133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return node
        
        oldToNew = {}
        
        def dfs(node):
            
            if node in oldToNew:
                return oldToNew[node]
            
            copy = Node(node.val)
            oldToNew[node] = copy
            
            for nbr in node.neighbors:
                copy.neighbors.append(dfs(nbr))
            
            return copy
        
        return dfs(node)

File: 133-clone-graph/133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public static Node dfs(Node node, Map<Node, Node> oldToNew){
        
        if(oldToNew.containsKey(node)){
            return oldToNew.get(node);
        }
        
        Node copy = new Node(node.val);
        oldToNew.put(node, copy);
        
        for(Node nbr : node.neighbors){
            copy.neighbors.add(dfs(nbr, oldToNew));
        }
        
        return copy;
    }
    
    public Node cloneGraph(Node node) {
        
        if(node == null) return node;
        
        Map<Node, Node> oldToNew = new HashMap<>();
        
        return dfs(node, oldToNew);
    }
}